synapse/lib/hiveauth.py

```diff
@@ -1270,6 +1400,9 @@ async def setPasswd(self, passwd, nexs=True):
             shadow = await s_passwd.getShadowV2(passwd=passwd)
         else:
             raise s_exc.BadArg(mesg='Password must be a string')
+
+        await self._checkPasswdPolicy(passwd, shadow)
```

Comment:
technically we can defer the generation of the new shadow until password is valid per the policy. saves a little bit of time in the sad path case.

Reply:
I did it this way because we store the new shadow in the `policy:previous` info to check password history. It does take a little longer to generate the shadow if we're not using it bue it makes all the password check (and any state tracking) all nicely contained in `_checkPasswdPolicy()`.